akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -699,18 +790,63 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
-  final def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
+  final protected def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
     new AsyncCallback[T] {
       override def invoke(event: T): Unit =
         interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
     }
   }
 
+  // TODO rethink if it's worth it; move the Updater to an object; etc
+  // via atomics since users may do something weird - try to Future { init } or something...
+  private final val _setStageActorRef = AtomicReferenceFieldUpdater.newUpdater(classOf[GraphStageLogic], classOf[StageActorRef], "_stageActorRefDoNotCallMeDirectly")
```

Comment:
I tried to move the updater to an object... and, wat - seems to be impossible due to the Updater being confused about some private volatile field? (even if I make the `_stageActorRefDoNotCallMeDirectly ` public, so it's not that which it is complaining about). Is it impossible to use those in Scala from a companion into the class for some reason?

// to be discussed anyway, if we need this or not. Also I was trying to make it possible to not be forced into using `implicit var` like @drewhk has to now in the TcpStages.